### ISSUE TITLE
[java] ArrayIsStoredDirectly not triggered from variadic functions

### ISSUE THREAD

**Issue by stephengroat** [pmd/pmd#963]
**Affects PMD Version:** 6.0.1

**Rule:** ArrayIsStoredDirectly

**Description:**

`ArrayIsStoredDirectly` does not to seem to be triggered from variadic functions, which always produce an array from arguments

**Code Sample demonstrating the issue:**

```
public class Test {
  private String[] strings;
  public void setStrings(String... strings) {
    this.strings = strings;
  }
}
```

**Running PMD through:** *[Maven]*

**Comment by oowekyala** [comment]
This is definitely a false positive. My guess is that we should make the rule use type resolution to solve that uniformly and simplify the rule, though it may lead to some false positives if the auxclasspath is not set. @jsotuyod should we do that? That fix would also depend on #912's [f2ba4e442c0](https://github.com/pmd/pmd/pull/912/commits/f2ba4e442c07ce3266c4b77ffb60683624b385be)

Btw it's a different problem, but the rule is also oblivious to assignments in an expression, eg
```java
public void setStrings(String[] strings) {
    System.err.println("strings is assigned to " + (this.strings = strings));
}
```

**Comment by jsotuyod** [comment]
@oowekyala a simpler fix would be to check for `isVarargs()` along when checking for `isArray()`

https://github.com/pmd/pmd/blob/0967d5027539d8528deb025d6b6d0ecece41d8f9/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/ArrayIsStoredDirectlyRule.java#L159

Non the less, I'm perfectly comfortable pushing for type resolution, Java 10 will actually need it (#743), and we have worked a lot towards reporting missing classes in auxclasspath (#194)
